packages/store/onboardingSlice/thunks.js

```diff
@@ -0,0 +1,141 @@
+import { getDemoDeviceAddress } from '@northern.tech/utils/helpers';
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import Cookies from 'universal-cookie';
+
+import { actions, sliceName } from '.';
+
+const cookies = new Cookies();
+
+export const applyOnboardingFallbacks = progress => {
+  const step = onboardingSteps[progress];
+  if (step && step.fallbackStep) {
+    return step.fallbackStep;
+  }
+  return progress;
+};
+
+const determineProgress = (acceptedDevices, pendingDevices, releases, pastDeployments) => {
+  const steps = onboardingSteps;
+  let progress = -1;
+  progress = pendingDevices.length > 1 ? steps.findIndex(step => step === onboardingStepNames.DEVICES_PENDING_ACCEPTING_ONBOARDING) : progress;
+  progress = acceptedDevices.length >= 1 ? steps.findIndex(step => step === onboardingStepNames.DEVICES_ACCEPTED_ONBOARDING) : progress;
+  progress =
+    acceptedDevices.length > 1 && releases.length > 1 && pastDeployments.length > 1
+      ? steps.findIndex(step => step === onboardingStepNames.DEPLOYMENTS_PAST_COMPLETED)
+      : progress;
+  return steps[progress];
+};
+
+const deductOnboardingState = ({ devicesById, devicesByStatus, onboardingState, pastDeployments, releases, userCapabilities, userId }) => {
+  const { canDeploy, canManageDevices, canReadDeployments, canReadDevices, canReadReleases, canUploadReleases } = userCapabilities;
+  const userCookie = cookies.get(`${userId}-onboarded`);
+  const acceptedDevices = devicesByStatus[DEVICE_STATES.accepted].deviceIds;
+  const pendingDevices = devicesByStatus[DEVICE_STATES.pending].deviceIds;
+  let deviceType = onboardingState.deviceType ?? [];
+  deviceType =
+    !deviceType.length && acceptedDevices.length && devicesById[acceptedDevices[0]].hasOwnProperty('attributes')
+      ? devicesById[acceptedDevices[0]].attributes.device_type
+      : deviceType;
+  const progress = applyOnboardingFallbacks(onboardingState.progress || determineProgress(acceptedDevices, pendingDevices, releases, pastDeployments));
+  return {
+    ...onboardingState,
+    complete: !!(
+      Boolean(userCookie) ||
+      onboardingState.complete ||
+      (acceptedDevices.length > 1 && pendingDevices.length > 0 && releases.length > 1 && pastDeployments.length > 1) ||
+      (acceptedDevices.length >= 1 && releases.length >= 2 && pastDeployments.length > 2) ||
+      (acceptedDevices.length >= 1 && pendingDevices.length > 0 && releases.length >= 2 && pastDeployments.length >= 2) ||
+      onboardingSteps.findIndex(step => step === progress) >= onboardingSteps.length - 1 ||
+      onboardingState.disable ||
+      ![canDeploy, canManageDevices, canReadDeployments, canReadDevices, canReadReleases, canUploadReleases].every(i => i)
+    ),
+    showTips: onboardingState.showTips != null ? onboardingState.showTips : true,
+    deviceType,
+    approach: onboardingState.approach || (deviceType.some(type => type.startsWith('qemu')) ? 'virtual' : 'physical'),
```

Comment:
'virtual' and 'physical' might be moved into constants, I see them in different places

Reply:
will do 👌